src/sabre_swap/mod.rs

```diff
@@ -0,0 +1,198 @@
+/// Candidate swaps are sorted so SWAP(i,j) and SWAP(j,i) are not duplicated.
+fn obtain_swaps(
+    front_layer: &EdgeList,
+    neighbors: &NeighborTable,
+    layout: &NLayout,
+) -> HashSet<[usize; 2]> {
+    // This will likely under allocate as it's a function of the number of
+    // neighbors for the qubits in the layer too, but this is basically a
+    // minimum allocation assuming each qubit has only 1 unique neighbor
+    let mut candidate_swaps: HashSet<[usize; 2]> =
+        HashSet::with_capacity(2 * front_layer.edges.len());
+    for node in &front_layer.edges {
+        for v in node {
+            let physical = layout.logic_to_phys[*v];
+            for neighbor in &neighbors.neighbors[physical] {
+                let virtual_neighbor = layout.phys_to_logic[*neighbor];
+                let swap: [usize; 2] = if &virtual_neighbor > v {
+                    [*v, virtual_neighbor]
+                } else {
+                    [virtual_neighbor, *v]
+                };
+                candidate_swaps.insert(swap);
+            }
+        }
+    }
+    candidate_swaps
+}
+
+/// Run the sabre heuristic scoring
+///
+/// Args:
+///     layers (EdgeList): The input layer edge list to score and find the
+///         best swaps
+///     layout (NLayout): The current layout
+///     neighbor_table (NeighborTable): The table of neighbors for each node
+///         in the coupling graph
+///     extended_set (EdgeList): The extended set
+///     distance_matrix (ndarray): The 2D array distance matrix for the coupling
+///         graph
+///     qubits_decay (QubitsDecay): The current qubit decay factors for
+///     heuristic (Heuristic): The chosen heuristic method to use
+/// Returns:
+///     ndarray: A 2d array of the best swap candidates all with the minimum score
+#[pyfunction]
+pub fn sabre_score_heuristic(
+    layer: EdgeList,
+    layout: &mut NLayout,
+    neighbor_table: &NeighborTable,
+    extended_set: EdgeList,
+    distance_matrix: PyReadonlyArray2<f64>,
+    qubits_decay: QubitsDecay,
+    heuristic: &Heuristic,
+    rng: &mut SabreRng,
+) -> [usize; 2] {
+    let dist = distance_matrix.as_array();
+    let candidate_swaps = obtain_swaps(&layer, neighbor_table, layout);
+    let mut min_score = f64::MAX;
+    let mut best_swaps: Vec<[usize; 2]> = Vec::new();
+    for swap_qubits in candidate_swaps {
+        layout.swap_logical(swap_qubits[0], swap_qubits[1]);
+        let score = score_heuristic(
+            heuristic,
+            &layer.edges,
+            &extended_set.edges,
+            layout,
+            &swap_qubits,
+            &dist,
+            &qubits_decay.decay,
+        );
+        if score < min_score {
+            min_score = score;
+            best_swaps.clear();
+            best_swaps.push(swap_qubits);
+        } else if score == min_score {
+            best_swaps.push(swap_qubits);
+        }
+        layout.swap_logical(swap_qubits[0], swap_qubits[1]);
+    }
+    best_swaps.par_sort();
```

Comment:
Should this respect `getenv_use_multiple_threads()`?

Reply:
Fixed in: https://github.com/Qiskit/qiskit-terra/pull/7977/commits/3ce5ff853e2d83d44f33b10acafd7bdd364c5e83